pkg/util/stringencoding/string_encoding.go

```diff
@@ -109,11 +108,25 @@ func EncodeEscapedChar(
 			// Escape non-printable characters.
 			buf.Write(HexMap[currentByte])
 		}
-	} else if ln == 2 {
-		// For multi-byte runes, print them based on their width.
-		fmt.Fprintf(buf, `\u%04X`, currentRune)
 	} else {
-		fmt.Fprintf(buf, `\U%08X`, currentRune)
+		writeMultibyteRuneAsHex(buf, currentRune, ln)
+	}
+}
+
+const uppercaseHex = `0123456789ABCDEF`
+
+// writeMultibyteRuneAsHex is equivalent to either
+// fmt.FPrintf(`\u%04X`) or fmt.FPrintf(`\U%08X`).
+// We can't quite just use strconv since we need uppercase hex.
+func writeMultibyteRuneAsHex(buf *bytes.Buffer, r rune, ln int) {
+	if ln == 2 {
+		buf.WriteString(`\u0000`)
+	} else {
+		buf.WriteString(`\U00000000`)
+	}
```

Comment:
I wonder if simply having
```
var scratch [8]byte
```
writing to that, and just buf.Append(scratch[:ln]) might be better/faster.

Reply:
I'd be surprised if it was faster, but I'll try that out later. My version writes to the same place twice, but yours writes the same bytes twice to different places.

Reply:
Might not be faster... But... Wirting directly to internal Bytes() buffer -- sometimes feels a bit magical.
Scratch I think is clearer.  And fwiw, it's used a lot in Appent* functions;